### PR TITLE
sdk support push to graphite

### PR DESCRIPTION
**Pull request by zhijian-pro**

### CODE REVIEW

File: sdk/java/libjfs/bridge.go [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+	vec, err := expfmt.ExtractSamples(&expfmt.DecodeOptions{
+		Timestamp: now,
+	}, mfs...)
+	if err != nil {
+		return err
+	}
+
+	buf := bufio.NewWriter(w)
+	for _, s := range vec {
+		if prefix != "" {
```
Comment (zhijian-pro) [inline sdk/java/libjfs/bridge.go:219]: The original logic here is to still write a point when the prefix is empty. I think this is a bug, so I modified it here. When prefix is empty, dots should not be written
```suggestion
		if prefix != "" {
```

**Review by xiaogaozi — APPROVED** [review]

File: sdk/java/libjfs/main.go [hunk 2]
```diff
@@ -306,6 +308,70 @@ func createStorage(format meta.Format) (object.ObjectStorage, error) {
...
+			UseTags:       true,
+			Timeout:       2 * time.Second,
+			ErrorHandling: ContinueOnError,
+			Logger:        logger,
+			CommonLabels:  commonLabels,
+		})
+		if err != nil {
+			logger.Warnf("NewBridge error:%s", err)
+		}
+		bridges = append(bridges, bridge)
```
Comment (davies) [inline sdk/java/libjfs/main.go:325]: bridge could be nil

File: sdk/java/libjfs/main.go [hunk 3]
```diff
@@ -306,6 +308,70 @@ func createStorage(format meta.Format) (object.ObjectStorage, error) {
 	return object.WithPrefix(blob, format.Name+"/"), nil
 }
 
+func pushMetrics(graphite, pushGatewayAddr, pushAuth string, graphiteInterVal, pushGatewayInterVal int, registry *prometheus.Registry, commonLabels map[string]string) {
```
Comment (davies) [inline sdk/java/libjfs/main.go:311]: it's better split it into two function, one for push gateway, one for graphite

File: sdk/java/libjfs/main.go [hunk 4]
```diff
@@ -306,6 +308,70 @@ func createStorage(format meta.Format) (object.ObjectStorage, error) {
...
+		if pushAuth != "" {
+			if strings.Contains(pushAuth, ":") {
+				parts := strings.Split(pushAuth, ":")
+				pusher.BasicAuth(parts[0], parts[1])
+			}
+		}
+		pushers = append(pushers, pusher)
+	}
+
+	once.Do(func() {
```
Comment (davies) [inline sdk/java/libjfs/main.go:342]: We could have one goroutine  for each pusher

Comment (davies) [reply sdk/java/libjfs/main.go:342]: Or one for all pushers and bridges

File: sdk/java/libjfs/bridge.go (hunk 1 above)
Comment (zhijian-pro) [reply sdk/java/libjfs/bridge.go:219]: This is indeed a bug, upstream has been merged.https://github.com/prometheus/client_golang/pull/1005

**Comment by zhijian-pro** [comment]
**graphite:**
image

**pushgateway:**
image

**Pull request merged**